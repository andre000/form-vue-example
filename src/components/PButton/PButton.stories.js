import { action } from '@storybook/addon-actions'
import { withKnobs, text, boolean } from '@storybook/addon-knobs'
import PButton from './PButton.vue'

export default {
  component: PButton,
  title: 'PButton',
  decorator: [withKnobs]
}

export const withText = () => ({
  components: { PButton },
  props: {
    text: {
      default: text('Text', 'Default Button')
    }
  },
  template: '<p-button @click="action">{{ text }}</p-button>',
  methods: { action: action('clicked') }
})

export const withPrimaryColor = () => ({
  components: { PButton },
  props: {
    disabled: {
      default: boolean('Disabled', false)
    }
  },
  template:
    '<p-button @click="action" :disabled="disabled" color="primary">Button</p-button>',
  methods: { action: action('clicked') }
})

export const withSecondaryColor = () => ({
  components: { PButton },
  props: {
    disabled: {
      default: boolean('Disabled', false)
    }
  },
  template:
    '<p-button @click="action" :disabled="disabled" color="secondary">Button</p-button>',
  methods: { action: action('clicked') }
})
