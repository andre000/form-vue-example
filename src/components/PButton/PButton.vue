<template>
  <button class="button" :class="`button--${color}`">
    <slot />
  </button>
</template>

<script>
export default {
  name: 'PButton',
  props: {
    color: {
      type: String,
      default: 'default',
      validator: v => ['default', 'primary', 'secondary'].includes(v)
    }
  }
}
</script>

<style lang="scss" scoped>
@import '@/assets/scss/variables.scss';

.button {
  width: 200px;
  height: 60px;
  border-radius: 30px;
  outline: none;
  border: none;
  font: 900 20px/26px Roboto;
  text-transform: uppercase;
  cursor: pointer;
  transition: filter 0.3s ease;

  &:hover {
    filter: drop-shadow(2px 4px 6px $shadow-color);
  }
  &:disabled {
    filter: contrast(0.5) grayscale(0.6);
    cursor: not-allowed;
  }

  &.button--default {
    background: #eee;
  }

  &.button--primary {
    background: $primary;
    color: #fff;
  }

  &.button--secondary {
    background: $secondary;
    color: $primary-text;
  }
}
</style>
