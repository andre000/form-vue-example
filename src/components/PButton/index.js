export { default as PButton } from './PButton.vue'
