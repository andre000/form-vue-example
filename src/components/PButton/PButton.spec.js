import { shallowMount, createLocalVue } from '@vue/test-utils'

import PButton from './PButton'

const localVue = createLocalVue()

describe('PButton component', () => {
  let wrapper
  beforeEach(() => {
    wrapper = shallowMount(PButton, {
      localVue
    })
  })

  it('should match the snapshot', () =>
    expect(wrapper.html()).toMatchSnapshot())

  it('should have button class', () =>
    expect(wrapper.find('.button').exists()).toBe(true))

  it('should have button--default when the color prop is not defined', () => {
    expect(wrapper.find('.button--default').exists()).toBe(true)
  })

  it('should change class along with the color prop', async () => {
    await wrapper.setProps({ color: 'primary' })
    expect(wrapper.find('.button--primary').exists()).toBe(true)
    await wrapper.setProps({ color: 'secondary' })
    expect(wrapper.find('.button--secondary').exists()).toBe(true)
  })

  it('should throw an error when using a not valid color prop', async () => {
    expect(wrapper.vm.$options.props.color.validator('foo')).toBe(false)
  })
})
