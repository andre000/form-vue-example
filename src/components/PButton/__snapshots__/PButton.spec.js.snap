// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PButton component should match the snapshot 1`] = `<button class="button button--default"></button>`;
